File: tools/release/build_maven_artifact_manifest.py
#!/usr/bin/env python3
"""Build a manifest for Oliphaunt tarball Maven artifact publications."""

from __future__ import annotations

import argparse
import sys
from pathlib import Path
from typing import NoReturn

import extension_artifact_targets
import product_metadata


ROOT = Path(__file__).resolve().parents[2]


def fail(message: str) -> NoReturn:
    print(f"build_maven_artifact_manifest.py: {message}", file=sys.stderr)
    raise SystemExit(1)


def repo_path(value: str) -> Path:
    path = Path(value)
    if not path.is_absolute():
        path = ROOT / path
    return path


def require_file(path: Path, label: str) -> Path:
    if not path.is_file():
        fail(f"missing {label}: {path.relative_to(ROOT)}")
    return path


def tsv_row(
    *,
    group_id: str,
    artifact_id: str,
    version: str,
    file: Path,
    name: str,
    description: str,
) -> str:
    values = [group_id, artifact_id, version, str(file.relative_to(ROOT)), name, description]
    if any("\t" in value or "\n" in value for value in values):
        fail(f"Maven artifact manifest value contains a tab or newline: {values}")
    return "\t".join(values)


RUNTIME_MAVEN_ARTIFACTS = {
    "liboliphaunt-runtime-resources": {
        "filename": "liboliphaunt-{version}-runtime-resources.tar.gz",
        "name": "Oliphaunt runtime resources",
        "description": "Package-managed Oliphaunt PostgreSQL runtime resources for Android app builds.",
    },
    "oliphaunt-icu": {
        "filename": "liboliphaunt-{version}-icu-data.tar.gz",
        "name": "Oliphaunt ICU data",
        "description": "Package-managed optional ICU data files for Oliphaunt app builds.",
    },
    "liboliphaunt-android-arm64-v8a": {
        "filename": "liboliphaunt-{version}-android-arm64-v8a.tar.gz",
        "name": "Oliphaunt Android runtime arm64-v8a",
        "description": "Package-managed liboliphaunt Android runtime for arm64-v8a app builds.",
    },
    "liboliphaunt-android-x86_64": {
        "filename": "liboliphaunt-{version}-android-x86_64.tar.gz",
        "name": "Oliphaunt Android runtime x86_64",
        "description": "Package-managed liboliphaunt Android runtime for x86_64 app builds.",
    },
}


def split_maven_coordinate(coordinate: str) -> tuple[str, str]:
    group_id, separator, artifact_id = coordinate.partition(":")
    if not separator or not group_id or not artifact_id:
        fail(f"invalid Maven coordinate {coordinate!r}; expected group:artifact")
    return group_id, artifact_id


def runtime_rows(asset_root: Path) -> list[str]:
    version = product_metadata.read_current_version("liboliphaunt-native")
    rows = []
    for coordinate in product_metadata.registry_package_names("liboliphaunt-native", "maven"):
        group_id, artifact_id = split_maven_coordinate(coordinate)
        if group_id != "dev.oliphaunt.runtime":
            fail(f"liboliphaunt-native Maven artifact {coordinate} must use dev.oliphaunt.runtime")
        artifact = RUNTIME_MAVEN_ARTIFACTS.get(artifact_id)
        if artifact is None:
            fail(f"liboliphaunt-native Maven artifact {coordinate} has no release asset mapping")
        filename = artifact["filename"].format(version=version)
        rows.append(
            tsv_row(
                group_id=group_id,
                artifact_id=artifact_id,
                version=version,
                file=require_file(asset_root / filename, artifact_id),
                name=artifact["name"],
                description=artifact["description"],
            )
        )
    return rows


def extension_rows(extension_root: Path, selected_products: list[str]) -> list[str]:
    products = selected_products or [
        product
        for product in product_metadata.product_ids()
        if product_metadata.product_config(product).get("kind") == "exact-extension-artifact"
    ]
    rows: list[str] = []
    for product in sorted(products):
        config = product_metadata.product_config(product)
        if config.get("kind") != "exact-extension-artifact":
            fail(f"{product} is not an exact-extension-artifact product")
        sql_name = config.get("extension_sql_name")
        if not isinstance(sql_name, str) or not sql_name:
            fail(f"{product} release metadata must declare extension_sql_name")
        version = product_metadata.read_current_version(product)
        product_root = extension_root / product / "release-assets"
        targets = extension_artifact_targets.published_android_maven_targets(product)
        if not targets:
            fail(f"{product} has no published Android Maven extension targets")
        for target in targets:
            filename = f"{product}-{version}-native-{target.target}-runtime.tar.gz"
            rows.append(
                tsv_row(
                    group_id="dev.oliphaunt.extensions",
                    artifact_id=f"{product}-{target.target}",
                    version=version,
                    file=require_file(product_root / filename, f"{product} {target.target} Maven artifact"),
                    name=f"Oliphaunt extension {sql_name} {target.target}",
                    description=f"Package-managed Oliphaunt Android runtime and static-link artifacts for the {sql_name} PostgreSQL extension on {target.target}.",
                )
            )
    return rows


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", required=True, help="TSV manifest path to write")
    parser.add_argument(
        "--runtime-asset-root",
        default="target/liboliphaunt/release-assets",
        help="Directory containing liboliphaunt runtime release assets",
    )
    parser.add_argument(
        "--extension-artifact-root",
        default="target/extension-artifacts",
        help="Directory containing staged exact-extension package artifacts",
    )
    parser.add_argument("--runtime", action="store_true", help="include base liboliphaunt Android runtime artifacts")
    parser.add_argument("--extensions", action="store_true", help="include Android exact-extension artifacts")
    parser.add_argument("--extension-product", action="append", default=[], help="exact-extension product to include")
    args = parser.parse_args()

    include_runtime = args.runtime or not args.extensions
    include_extensions = args.extensions or bool(args.extension_product)
    rows: list[str] = []
    if include_runtime:
        rows.extend(runtime_rows(repo_path(args.runtime_asset_root)))
    if include_extensions:
        rows.extend(extension_rows(repo_path(args.extension_artifact_root), args.extension_product))
    if not rows:
        fail("manifest would be empty")

    output = repo_path(args.output)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text("\n".join(rows) + "\n", encoding="utf-8")
    print(f"Wrote {len(rows)} Maven artifact publication row(s) to {output.relative_to(ROOT)}")


if __name__ == "__main__":
    main()

File: tools/release/build_maven_artifact_manifest.mjs
#!/usr/bin/env bun
import fs from "node:fs/promises";
import path from "node:path";

import { currentVersion } from "./product-version.mjs";
import {
  allArtifactTargets,
  exactExtensionProducts as logicalExactExtensionProducts,
  extensionArtifactProductRoot,
  extensionArtifactTargets as releaseExtensionArtifactTargets,
  extensionMetadata,
  extensionReleaseProduct,
  extensionReleaseVersion,
  extensionSqlNames,
  registryPackageRows,
} from "./release-artifact-targets.mjs";
import {
  assertReleaseNoticesInEntries,
  releaseProfileMavenLicenses,
  releaseProfilePackageLicense,
} from "./release-notices.mjs";
import {
  assertExtensionUpstreamLicensesInEntries,
  extensionMavenLicenses,
  extensionRegistryLicense,
} from "./extension-upstream-licenses.mjs";
import { readPortableArchiveEntries } from "./portable-archive.mjs";

const ROOT = path.resolve(import.meta.dir, "../..");
const PREFIX = "build_maven_artifact_manifest.mjs";

function compareText(left, right) {
  return left < right ? -1 : left > right ? 1 : 0;
}

function fail(message) {
  console.error(`${PREFIX}: ${message}`);
  process.exit(1);
}

function rel(file) {
  return path.relative(ROOT, file).split(path.sep).join("/");
}

function canonicalMavenPayloadPrefix(file, entries) {
  if (entries.has("LICENSE")) return "";

  const roots = new Set([...entries.keys()].map((member) => member.split("/", 1)[0]));
  const expected = path.basename(file).slice(0, -".tar.gz".length);
  if (roots.size !== 1 || !roots.has(expected)) {
    throw new Error(
      `${path.basename(file)} must stage release notices at the archive root or beneath its `
      + `canonical single archive root ${expected}`,
    );
  }
  return expected;
}

function assertMavenPayloadLegal(file, profile, sqlNames = []) {
  try {
    const entries = readPortableArchiveEntries(file);
    const prefix = canonicalMavenPayloadPrefix(file, entries);
    assertReleaseNoticesInEntries(entries, { profile, prefix, label: path.basename(file) });
    if (sqlNames.length > 0) {
      // Native singleton payloads install upstream notices with runtime files
      // below files/, while aggregate bundles stage their combined legal tree
      // directly below the canonical archive root.
      const upstreamPrefix = prefix === "" ? "files" : prefix;
      assertExtensionUpstreamLicensesInEntries(sqlNames, entries, { prefix: upstreamPrefix });
    }
  } catch (error) {
    fail(`${rel(file)} failed Maven payload legal closure: ${error instanceof Error ? error.message : String(error)}`);
  }
  return file;
}

function repoPath(value) {
  return path.isAbsolute(value) ? value : path.join(ROOT, value);
}

function nativeRuntimeArtifactTargets(version) {
  return allArtifactTargets({
    product: "liboliphaunt-native",
    publishedOnly: true,
  }, PREFIX)
    .filter((target) => target.surfaces.includes("maven"))
    .map((target) => ({
      ...target,
      asset: target.asset.replaceAll("{version}", version),
    }))
    .sort((left, right) => compareText(left.id, right.id));
}

function runtimeMavenArtifactId(target) {
  if (target.kind === "runtime-resources") {
    return "liboliphaunt-runtime-resources";
  }
  if (target.kind === "icu-data") {
    return "oliphaunt-icu";
  }
  if (target.kind === "native-runtime" && target.target.startsWith("android-")) {
    return `liboliphaunt-${target.target}`;
  }
  return undefined;
}

function runtimeMavenArtifactMetadata(target) {
  if (target.kind === "runtime-resources") {
    return {
      name: "Oliphaunt runtime resources",
      description: "Package-managed Oliphaunt PostgreSQL runtime resources for Android app builds.",
      licenseProfile: "native-runtime-resources",
    };
  }
  if (target.kind === "icu-data") {
    return {
      name: "Oliphaunt ICU data",
      description: "Package-managed optional ICU data files for Oliphaunt app builds.",
      licenseProfile: "native-icu-data",
    };
  }
  if (target.kind === "native-runtime" && target.target.startsWith("android-")) {
    const abi = target.target.slice("android-".length);
    return {
      name: `Oliphaunt Android runtime ${abi}`,
      description: `Package-managed liboliphaunt Android runtime for ${abi} app builds.`,
      licenseProfile: "native-runtime",
    };
  }
  fail(`unsupported liboliphaunt-native Maven artifact target ${target.id}`);
}

function runtimeMavenArtifacts(version) {
  const artifacts = new Map();
  for (const target of nativeRuntimeArtifactTargets(version)) {
    const artifactId = runtimeMavenArtifactId(target);
    if (artifactId === undefined) {
      continue;
    }
    if (artifacts.has(artifactId)) {
      fail(`duplicate liboliphaunt-native Maven artifact mapping for ${artifactId}`);
    }
    artifacts.set(artifactId, {
      filename: target.asset,
      ...runtimeMavenArtifactMetadata(target),
    });
  }
  if (artifacts.size === 0) {
    fail("liboliphaunt-native artifact targets did not produce any Maven runtime artifacts");
  }
  return artifacts;
}

function splitMavenCoordinate(coordinate) {
  const separator = coordinate.indexOf(":");
  if (separator <= 0 || separator === coordinate.length - 1) {
    fail(`invalid Maven coordinate ${JSON.stringify(coordinate)}; expected group:artifact`);
  }
  return [coordinate.slice(0, separator), coordinate.slice(separator + 1)];
}

async function requireFile(file, label) {
  try {
    const stat = await fs.stat(file);
    if (stat.isFile()) {
      return file;
    }
  } catch {
    // Fall through to the shared diagnostic below.
  }
  fail(`missing ${label}: ${rel(file)}`);
}

function tsvRow({
  groupId,
  artifactId,
  version,
  file,
  name,
  description,
  runtimeProduct = "",
  runtimeVersion = "",
  licenseSpdx,
  licenses,
}) {
  if (typeof licenseSpdx !== "string" || !licenseSpdx) fail(`Maven artifact ${groupId}:${artifactId} has no package SPDX expression`);
  if (!Array.isArray(licenses) || licenses.length === 0) fail(`Maven artifact ${groupId}:${artifactId} has no structured license entries`);
  const values = [
    groupId,
    artifactId,
    version,
    rel(file),
    name,
    description,
    runtimeProduct,
    runtimeVersion,
    licenseSpdx,
    JSON.stringify(licenses),
  ];
  if (values.some((value) => value.includes("\t") || value.includes("\n"))) {
    fail(`Maven artifact manifest value contains a tab or newline: ${JSON.stringify(values)}`);
  }
  return values.join("\t");
}

async function runtimeRows(assetRoot) {
  const version = await currentVersion("liboliphaunt-native");
  const artifacts = runtimeMavenArtifacts(version);
  const rows = [];
  for (const coordinate of registryPackageRows({
    product: "liboliphaunt-native",
    packageKind: "maven",
  }, PREFIX).map((row) => row.packageName).filter((name) => name.startsWith("dev.oliphaunt.runtime:"))) {
    const [groupId, artifactId] = splitMavenCoordinate(coordinate);
    if (groupId !== "dev.oliphaunt.runtime") {
      fail(`liboliphaunt-native Maven artifact ${coordinate} must use dev.oliphaunt.runtime`);
    }
    const artifact = artifacts.get(artifactId);
    if (artifact === undefined) {
      fail(`liboliphaunt-native Maven artifact ${coordinate} has no release asset mapping`);
    }
    const file = await requireFile(path.join(assetRoot, artifact.filename), artifactId);
    assertMavenPayloadLegal(file, artifact.licenseProfile);
    rows.push(
      tsvRow({
        groupId,
        artifactId,
        version,
        file,
        name: artifact.name,
        description: artifact.description,
        licenseSpdx: releaseProfilePackageLicense(artifact.licenseProfile).spdx,
        licenses: releaseProfileMavenLicenses(artifact.licenseProfile, {
          product: "liboliphaunt-native",
          version,
        }),
      }),
    );
  }
  return rows;
}

async function extensionRows(extensionRoot, selectedProducts) {
  const products = selectedProducts.length > 0
    ? selectedProducts
    : logicalExactExtensionProducts(PREFIX);
  const rows = [];
  for (const product of [...products].sort()) {
    const sqlNames = extensionSqlNames(product, PREFIX);
    const version = extensionReleaseVersion(product, "native", PREFIX);
    const registryLicense = extensionRegistryLicense(product, sqlNames);
    const compatibility = extensionMetadata(product, PREFIX).compatibility;
    const releaseProduct = extensionReleaseProduct(product, "native", PREFIX);
    const runtimeProduct = compatibility.nativeRuntimeProduct;
    const runtimeVersion = compatibility.nativeRuntimeVersion;
    if (typeof runtimeProduct !== "string" || !runtimeProduct || typeof runtimeVersion !== "string" || !runtimeVersion) {
      fail(`${product} must declare exact native runtime compatibility for Maven carriers`);
    }
    const currentRuntimeVersion = await currentVersion(runtimeProduct);
    if (runtimeVersion !== currentRuntimeVersion) {
      fail(`${product} native runtime compatibility ${runtimeVersion} does not match ${runtimeProduct}@${currentRuntimeVersion}`);
    }
    const productRoot = path.join(
      extensionArtifactProductRoot(product, "native", extensionRoot, PREFIX),
      "release-assets",
    );
    const targets = [...new Map(releaseExtensionArtifactTargets({
      product,
      family: "native",
      publishedOnly: true,
    }, PREFIX).filter((target) =>
      target.kind === "native-static-registry" && target.target.startsWith("android-"))
      .map((target) => [target.target, target])).values()];
    if (targets.length === 0) {
      fail(`${product} has no published Android Maven extension targets`);
    }
    const declaredCoordinates = new Set(
      registryPackageRows({ product: releaseProduct, packageKind: "maven" }, PREFIX)
        .map((row) => row.packageName)
        .filter((name) => name.startsWith(`dev.oliphaunt.extensions:${product}-`)),
    );
    for (const target of targets) {
      const coordinate = `dev.oliphaunt.extensions:${product}-${target.target}`;
      if (!declaredCoordinates.delete(coordinate)) {
        fail(`${product} release metadata is missing Maven carrier ${coordinate}`);
      }
      const filename = sqlNames.length > 1
        ? `${product}-${version}-native-${target.target}-bundle.tar.gz`
        : `${product}-${version}-native-${target.target}-runtime.tar.gz`;
      const memberLabel = sqlNames.length === 1
        ? `the ${sqlNames[0]} PostgreSQL extension`
        : `the PostgreSQL 18 contrib bundle (${sqlNames.length} exact extension members)`;
      const file = await requireFile(
        path.join(productRoot, filename),
        `${product} ${target.target} Maven artifact`,
      );
      const licenseProfile = product === "oliphaunt-extension-contrib-pg18"
        ? "contrib-native-openssl"
        : "external-native";
      assertMavenPayloadLegal(
        file,
        licenseProfile,
        product === "oliphaunt-extension-contrib-pg18" ? [] : sqlNames,
      );
      rows.push(
        tsvRow({
          groupId: "dev.oliphaunt.extensions",
          artifactId: `${product}-${target.target}`,
          version,
          file,
          name: `Oliphaunt ${sqlNames.length === 1 ? `extension ${sqlNames[0]}` : "PostgreSQL 18 contrib extensions"} ${target.target}`,
          description: `Package-managed Oliphaunt Android runtime and static-link artifacts for ${memberLabel} on ${target.target}.`,
          runtimeProduct,
          runtimeVersion,
          licenseSpdx: product === "oliphaunt-extension-contrib-pg18"
            ? releaseProfilePackageLicense(licenseProfile).spdx
            : registryLicense.packageSpdx,
          licenses: product === "oliphaunt-extension-contrib-pg18"
            ? releaseProfileMavenLicenses("contrib-native-openssl", { product, version })
            : extensionMavenLicenses(product, sqlNames, { version }),
        }),
      );
    }
    if (declaredCoordinates.size > 0) {
      fail(`${product} declares unexpected Maven carrier(s): ${[...declaredCoordinates].sort().join(", ")}`);
    }
  }
  return rows;
}

function valueArg(argv, index, name) {
  const value = argv[index + 1];
  if (value === undefined || value.startsWith("--")) {
    fail(`${name} requires a value`);
  }
  return value;
}

function parseArgs(argv) {
  const args = {
    output: undefined,
    runtimeAssetRoot: "target/liboliphaunt/release-assets",
    extensionArtifactRoot: "target/extension-artifacts",
    runtime: false,
    extensions: false,
    extensionProducts: [],
  };
  for (let index = 0; index < argv.length; ) {
    const arg = argv[index];
    if (arg === "--output") {
      args.output = valueArg(argv, index, arg);
      index += 2;
    } else if (arg === "--runtime-asset-root") {
      args.runtimeAssetRoot = valueArg(argv, index, arg);
      index += 2;
    } else if (arg === "--extension-artifact-root") {
      args.extensionArtifactRoot = valueArg(argv, index, arg);
      index += 2;
    } else if (arg === "--runtime") {
      args.runtime = true;
      index += 1;
    } else if (arg === "--extensions") {
      args.extensions = true;
      index += 1;
    } else if (arg === "--extension-product") {
      args.extensionProducts.push(valueArg(argv, index, arg));
      index += 2;
    } else {
      fail(`unknown argument: ${arg}`);
    }
  }
  if (!args.output) {
    fail("--output is required");
  }
  return args;
}

export async function buildMavenArtifactManifest(outputValue, {
  runtimeAssetRoot = "target/liboliphaunt/release-assets",
  extensionArtifactRoot = "target/extension-artifacts",
  runtime = false,
  extensions = false,
  extensionProducts = [],
} = {}) {
  const includeRuntime = runtime || !extensions;
  const includeExtensions = extensions || extensionProducts.length > 0;
  const rows = [];
  if (includeRuntime) {
    rows.push(...(await runtimeRows(repoPath(runtimeAssetRoot))));
  }
  if (includeExtensions) {
    rows.push(...(await extensionRows(repoPath(extensionArtifactRoot), extensionProducts)));
  }
  if (rows.length === 0) {
    fail("manifest would be empty");
  }
  const output = repoPath(outputValue);
  await fs.mkdir(path.dirname(output), { recursive: true });
  await fs.writeFile(output, `${rows.join("\n")}\n`, "utf8");
  console.log(`Wrote ${rows.length} Maven artifact publication row(s) to ${rel(output)}`);
  return output;
}

if (import.meta.main) {
  const args = parseArgs(Bun.argv.slice(2));
  await buildMavenArtifactManifest(args.output, args);
}
